tests/returntypes/test_vjp_new.py

```diff
@@ -0,0 +1,665 @@
+    #     def circuit(weights, features):
+    #         for i in range(nwires):
+    #            qml.RX(features[i], wires=i)
+    #             qml.RX(weights[i], wires=i)
+    #         return [qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))]
+
+    #     vanilla_numpy.random.seed(42)
+
+    #     ndata = 100
+    #     data = [vanilla_numpy.random.randn(nwires).astype("float32") for _ in range(ndata)]
+    #     label = [vanilla_numpy.random.choice([1, 0]).astype("int") for _ in range(ndata)]
+
+    #     loss = tf.losses.SparseCategoricalCrossentropy()
+
+    #     params = tf.Variable(vanilla_numpy.random.randn(nwires).astype("float32"), trainable=True)
+    #     with tf.GradientTape() as tape:
+    #         probs = [circuit(params, d) for d in data]
+    #         loss_value = loss(label, probs)
+
+    #    grads = tape.gradient(loss_value, [params])
+    #    assert len(grads) == 1
+
+    @pytest.mark.jax
+    @pytest.mark.slow
+    def test_jax(self, tol):
+        """Tests that the output of the VJP transform
+        can be differentiated using JAX."""
+        import jax
+        from jax import numpy as jnp
+
+        dev = qml.device("default.qubit.jax", wires=2)
+        params_np = np.array([0.543, -0.654], requires_grad=True)
+        params = jnp.array(params_np)
+
+        @partial(jax.jit, static_argnums=1)
+        def cost_fn(x):
+            with qml.tape.QuantumTape() as tape:
+                ansatz(x[0], x[1])
+            dy = jax.numpy.array([-1.0, 0.0, 0.0, 1.0])
+            tape.trainable_params = {0, 1}
+            tapes, fn = qml.gradients.vjp(tape, dy, param_shift)
+            vjp = fn(dev.batch_execute(tapes))
+            return vjp
+
+        res = cost_fn(params)
+        assert np.allclose(res, expected(params), atol=tol, rtol=0)
+
+        res = jax.jacobian(cost_fn, argnums=0)(params)
+        exp = qml.jacobian(expected)(params_np)
+        assert np.allclose(res, exp, atol=tol, rtol=0)
+
+
+class TestBatchVJP:
+    """Tests for the batch VJP function"""
+
+    def test_one_tape_no_trainable_parameters(self):
+        """A tape with no trainable parameters will simply return None"""
+        dev = qml.device("default.qubit", wires=2)
+
+        with qml.tape.QuantumTape() as tape1:
+            qml.RX(0.4, wires=0)
+            qml.CNOT(wires=[0, 1])
+            qml.expval(qml.PauliZ(0))
+
+        with qml.tape.QuantumTape() as tape2:
+            qml.RX(0.4, wires=0)
+            qml.RX(0.6, wires=0)
+            qml.CNOT(wires=[0, 1])
+            qml.expval(qml.PauliZ(0))
+
+        tape1.trainable_params = {}
+        tape2.trainable_params = {0, 1}
+
+        tapes = [tape1, tape2]
+        dys = [np.array(1.0), np.array(1.0)]
+
+        v_tapes, fn = qml.gradients.batch_vjp(tapes, dys, param_shift)
+        assert len(v_tapes) == 4
+
+        # Even though there are 3 parameters, only two contribute
+        # to the VJP, so only 2*2=4 quantum evals
```

Comment:
Should this be checked?

Reply:
It is done by `assert len(v_tapes) == 4`